### PR TITLE
chore(prisma): type discovery + fast builds + clean monorepo

### LINKED ISSUE #8443: Enable Type Discovery
## Problem

When working on the `prisma/prisma` repository, the different parts of code does not share types properly. This results in poor compilation times and the inability to follow types easily. At the moment, we compile all the types every time we build the project, even when working locally. This should only be necessary during CI.

By enabling automatic discovery of types throughout the prisma/prisma repository, development would be faster as, instead of having to remember or lookup type information, the IDE will automatically provide the information where it's needed.

## Suggested solution
Detailed is a suggested solution from @millsp :

- [x] Define a single typescript file that will be used for all of our packages and type-checks only
- [x] Define a single typescript file that will be used for all of our packages and build only
   - The type-check `tsconfig.json` file will extend the `tsconfig.build.json`
- [x] Define the base TypeScript rules that will apply to all our projects in the monorepo
- [x] Define TS type lookup paths to expose type information accross the monorepo
- [x] Provide each project with a `tsconfig.build.json` and a `tsconfig.json`
- [x] Integrate the new build files into our current build scripts for each package of the monorepo
- [x] Add a new `npm` build (call it `npm dev`) script for local dev that skips slow `tsc`
- [x] Similarly investigate why do we need to recompile all the packages `X` that depend `A` if we have changes in `A`. In a monorepo, this should not be necessary (at least at dev time).
     - Probably has to do with our bundling logic?
- [x] Do the same for `engines-wrapper`

## Additional context

This has been uncovered as a tech debt initiative and prioritized due to its high impact and relatively low complexity and estimated time for completion.

### PR DESCRIPTION
**Pull request by millsp**
It's all in the title :point_up: 

closes https://github.com/prisma/prisma/issues/8443

### CODE REVIEW

**Comment by millsp** [comment]
Looks like there is a flaky https://buildkite.com/prisma/prisma2-test/builds/8640#5d584656-8612-4d1e-b8f2-531d8cfde31d/229-3354

**Comment by millsp** [comment]
The binary is not fully downloaded before we execute the tests `BINARY_SIZE 200704` (bytes)

**Comment by millsp** [comment]
:tada: Solved thanks to @williamluke4 :tada:

**Comment by millsp** [comment]
thanks @williamluke4 :tada:

**Comment by Jolg42** [comment]
I merged my PR before that one and it looks like it created a conflict 🙈

**Comment by Jolg42** [comment]
I just tried it and so far it works like a charm 💚

File: packages/client/src/__tests__/types/$transaction/index.test-d.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  expectError(await prisma.$transaction([prisma.$disconnect()]))
+  expectError(
+    await prisma.$transaction([
+      new Promise((res) => res('You Shall Not Pass')),
+    ]),
+  )
+  expectError(await prisma.$transaction([5]))
+  expectError(await prisma.$transaction(['str']))
+  expectError(await prisma.$transaction([{}]))
+})()
```
Comment (Jolg42) [inline packages/client/src/__tests__/types/$transaction/index.test-d.ts:31]: Looks like an unrelated addition, but nice!

File: packages/debug/src/index.ts [hunk 2]
```diff
@@ -24,8 +24,6 @@ Debug.enable = (namespace: string): void => {
 
 Debug.enabled = (namespace: string): boolean => DebugNode.enabled(namespace)
 
-export declare type Debugger = DebugLib.Debugger
```
Comment (Jolg42) [inline packages/debug/src/index.ts:27]: Did we use this somewhere?

File: packages/sdk/src/utils/load.ts [hunk 3]
```diff
@@ -1,11 +1,16 @@
...
+ * This is a wrapper around `require`
  * This is to avoid eval and hide require away from bundlers
  */
 export function load<T>(id: string): T {
   try {
     return require(id) as T
   } catch (e) {
+    const { size } = fs.statSync(id)
+
+    console.log('BINARY_SIZE', size)
```
Comment (Jolg42) [inline packages/sdk/src/utils/load.ts:13]: Looks unrelated to types discovery, what is it for?

File: scripts/setup.ts [hunk 4]
```diff
@@ -84,7 +84,7 @@ has to point to the dev version you want to promote, for example 2.1.0-dev.123`)
           if (
             ['@prisma/migrate', '@prisma/integration-tests'].includes(pkgName)
           ) {
-            run(pkgDir, 'pnpm rebuild')
+            await run(pkgDir, 'pnpm rebuild')
```
Comment (Jolg42) [inline scripts/setup.ts:87]: nice catch

**Review by Jolg42 — APPROVED** [review]
Love it 💚

File: packages/sdk/src/utils/load.ts (hunk 3 above)
Comment (millsp) [reply packages/sdk/src/utils/load.ts:13]: Oh yes, that was a debug for tests not passing. Good catch!

File: packages/client/src/__tests__/types/$transaction/index.test-d.ts (hunk 1 above)
Comment (millsp) [reply packages/client/src/__tests__/types/$transaction/index.test-d.ts:31]: Thanks! That should be removed

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -18,16 +18,16 @@ on:
...
       - 'src/.buildkite/**'
-      - 'src/.dockerignore'
+      - '.dockerignore'
       - '*.bench.ts'
-      - 'src/scripts/ci/publish.ts'
-      - 'src/graphs/**'
+      - 'scripts/ci/publish.ts'
+      - 'graphs/**'
   pull_request:
     paths-ignore:
```
Comment (000gooo) [inline .github/workflows/test.yml:26]: Pull request

**Review by 000gooo — COMMENTED** [review]
_[]()_
